core/selection.js

```diff
@@ -578,29 +578,48 @@
 		};
 	} )();
 
-	// Handle left, right, delete and backspace keystrokes next to non-editable elements
-	// by faking selection on them.
+	// Handle left, right keystrokes next to non-editable elements
+	// by faking selection on them. Delete and backspace keystrokes can
+	// delete empty paragraphs between the widgets.
 	function getOnKeyDownListener( editor ) {
 		var keystrokes = { 37: 1, 39: 1, 8: 1, 46: 1 };
 
+		function isDeleteAction( keystroke ) {
+			return ( keystroke === 8 || keystroke === 46 );
+		}
+
+		function isEmptyElement( element ) {
+			var text = ( element.$.textContent === undefined ? element.$.innerText : element.$.textContent );
+			return text === '';
+		}
+
 		return function( evt ) {
 			var keystroke = evt.data.getKeystroke();
 
 			// Handle only left/right/del/bspace keys.
-			if ( !keystrokes[ keystroke ] )
+			if ( !keystrokes[ keystroke ] ) {
 				return;
+			}
 
 			var sel = editor.getSelection(),
 				ranges = sel.getRanges(),
-				range = ranges[ 0 ];
+				range = ranges[ 0 ],
+				startElement;
 
 			// Handle only single range and it has to be collapsed.
-			if ( ranges.length != 1 || !range.collapsed )
+			if ( !sel.isCollapsed() ) {
 				return;
+			}
 
 			var next = range[ keystroke < 38 ? 'getPreviousEditableNode' : 'getNextEditableNode' ]();
 
 			if ( next && next.type == CKEDITOR.NODE_ELEMENT && next.getAttribute( 'contenteditable' ) == 'false' ) {
+			//added if case to allow removal of empty paragraphs, see incident #1572
+				startElement = sel.getStartElement();
+				if ( startElement.getName() === 'p' && isDeleteAction( keystroke ) && isEmptyElement( startElement ) ) {
```

Comment:
Right now removing paragraphs works for, well, paragraphs. We can expand this to also work for other block elements, like headings (`<h1>, <h2>` etc.). I think replacing the `startElement.getName() === 'p'` condition with [`startElement.isBlockBoundary()`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_element.html#method-isBlockBoundary) should do.

If that works, please add similar unit tests for such cases.

Reply:
yep. It should work.